sha1_mb/sha1_mb_mgr_flush_avx2.asm

```diff
@@ -120,56 +120,91 @@ sha1_mb_mgr_flush_avx2:
 %endif
 
 	; use num_lanes_inuse to judge all lanes are empty
-	cmp	dword [state + _num_lanes_inuse], 0
-	jz	return_null
-
-	; find a lane with a non-null job
-	xor	idx, idx
-	cmp	qword [state + _ldata + 1 * _LANE_DATA_size + _job_in_lane], 0
-	cmovne	idx, [one]
-	cmp	qword [state + _ldata + 2 * _LANE_DATA_size + _job_in_lane], 0
-	cmovne	idx, [two]
-	cmp	qword [state + _ldata + 3 * _LANE_DATA_size + _job_in_lane], 0
-	cmovne	idx, [three]
-	cmp	qword [state + _ldata + 4 * _LANE_DATA_size + _job_in_lane], 0
-	cmovne	idx, [four]
-	cmp	qword [state + _ldata + 5 * _LANE_DATA_size + _job_in_lane], 0
-	cmovne	idx, [five]
-	cmp	qword [state + _ldata + 6 * _LANE_DATA_size + _job_in_lane], 0
-	cmovne	idx, [six]
-	cmp	qword [state + _ldata + 7 * _LANE_DATA_size + _job_in_lane], 0
-	cmovne	idx, [seven]
-
-	; copy idx to empty lanes
+	
+	lea     tmp, [state+  _ldata + 1 * _LANE_DATA_size + _job_in_lane]
+        xor     DWORD(idx), DWORD(idx)
+        cmp     DWORD(idx), dword [tmp + _num_lanes_inuse -(_ldata + 1 * _LANE_DATA_size + _job_in_lane)] ; idx=zero. offset from tmp
+        jz      return_null
+
+
+	vpxor   xmm0, xmm0,xmm0
+        vpcmpeqq ymm1,ymm0,[tmp]
+        vpcmpeqq ymm2,ymm0,[tmp+4*_LANE_DATA_size +_job_in_lane]
+        vpmovmskb DWORD(tmp), ymm2 ; high cmp mask
+        vpmovmskb DWORD(idx), ymm1 ; low cmp mask
+        shl     tmp, 32
+        or      idx, tmp
+        not     idx
+        tzcnt   idx, idx
+        shr     DWORD(idx), 3 ; divide by 8
+        add     DWORD(idx), 1 ;
+
 copy_lane_data:
-	mov	tmp, [state + _args + _data_ptr + 8*idx]
+	lea     tmp, [state + 8*idx]
+	mov	ebp, _args + _data_ptr
+        or      DWORD(tmp4), -1 ; 0xFFFFFFFF
```

Comment:
OR reg, IMM sign extends the IMM to the full register size so it will set the register to all ones whaterver contents the reg had, it's just a trick to save code size instead of using someting like xor reg,reg ; neg reg to initialize the register to 0xFFFFFFFF

Reply:
ah, sorry I thought this was an xor.  Try to keep the tab format the same as the existing file.  Some of the leading tabs have been replaced by spaces.